Abstract Justification Toolbar to it's own block-editor component

mkaz opened the pull request:
## Description

Move the toolbar created for the Navigation block to add justification controls, mainly space-between, to it's own component within the block-editor components. We want to add these controls additional to the Social Links block, but first need this PR to abstract it out and then can use in both spots without duplicating work.

## How has this been tested?

1. Create a Navigation block using the justify controls
2. Apply the PR and confirm the block still works
3. Confirm the justify controls still work
4. You can modify the style.scss in the justify-toolbar component and confirm that is where the CSS comes from.

## Types of changes

No functional changes, just code organization and clean up making the component reusable.

mkaz committed:
Initial abstraction of justify to own component

mkaz committed:
Move style to component, include properly and stub

- Updates style moving to component
- Fix exporting for component
- Initial stubbed documentation

mkaz committed:
Update API to better match an abstract component

mkaz committed:
Update docs and changelog

mkaz committed:
Auto generated docs change

jasmussen commented:
Cool work, looks and works as advertised:

<img width="847" alt="Screenshot 2021-01-25 at 08 55 17" src="https://user-images.githubusercontent.com/1204802/105676692-2c63de00-5eeb-11eb-9978-f210a7d46d13.png">

Code looks good and this is a nice enhancement. I CC'ed Nik for a quick sanity check as I know he's been looking at adjacent things.

One question, though, about this code:

```
.items-justified-left > ul {
	justify-content: flex-start;
}
```

I know this is existing code that has shipped for months and months, so it's probably not feasible to course correct. But what this code tells me is that you can't use justification controls unless the child elements are inside an unordered list.

Is there any way to change this to be less opinionated about what elements are being used? Could be fine as a followup if possible.

ntsekouras commented on packages/block-editor/src/components/justify-toolbar/index.js:
```diff
@@ -0,0 +1,63 @@
+			popoverProps={ popoverProps }
+			label={ __( 'Change items justification' ) }
+			isCollapsed
+			controls={ [
```
I think it would make sense to have these controls as default, but be able to override what controls we should show here. For example in `buttons` we don't have `space-between` and I'm not sure if there was a reason for this..

Or something like `allowedControls: [ 'left, 'right' ]? 🤔

mkaz replied:
Implemented in 37ae8ad

ntsekouras commented on packages/block-editor/src/components/justify-toolbar/index.js:
```diff
@@ -0,0 +1,63 @@
+} from '@wordpress/icons';
+import { __ } from '@wordpress/i18n';
+
+const navIcons = {
```
This should be renamed to something more generic.

mkaz replied:
:+1: Addressed in b68cf5d733

ntsekouras commented on packages/block-editor/src/components/justify-toolbar/style.scss:
```diff
@@ -0,0 +1,16 @@
+// Justification.
```
This is really specific to `Navigation` block structure, which applies the `.items-justified-left || right...etc` css classes to a parent container, which has a child `ul`. Most blocks I think apply the class to the parent element and have afterwards the children to be positioned.

This is tricky here as it will not work well for everything. I think we should cover the more generic case and change accordingly the individual blocks that have the less common case.

For example buttons will need just: `.items-justified-left { justify-content: flex-start }`

mkaz replied:
Addressed in discussion below.

ntsekouras reviewed (commented):
Thanks for working on this @mkaz !

Maybe by including an integration in another component as well, besides `Navigation`, will be better, as it will help identify more things or problems we have to solve.

mkaz committed:
Rename icons to generic, not in nav anymore

mkaz committed:
Pass foward isCollapsed prop

mkaz committed:
Add allowedControls prop to allow limiting controls shown

mkaz commented:
@ntsekouras @jasmussen So looking at the CSS we could probably tackle it a couple of different ways, but tricky for implementation due to existing blocks, classnames, and markup.

So, what I'm thinking is don't include any CSS in the controls at let the component using them implement as needed. This is how it works on navigation, the justify controls takes an `onChange` function which is then used to insert the classes where needed.

Looking at how I would implement the same controls with CSS for buttons is tricky sine the buttons component uses a different set of classnames, so changing either way will have some deprecation issues.

The downside is the controls aren't simple to drop-in and you automatically get all the classnames needed.

Thoughts?

jasmussen commented:
> So, what I'm thinking is don't include any CSS in the controls at let the component using them implement as needed. This is how it works on navigation, the justify controls takes an onChange function which is then used to insert the classes where needed.

This _seems_ right to me. There will likely be enough use cases for justification controls, that the markup can't always be the same anyway. But I will defer to Nik!

ntsekouras commented:
>So, what I'm thinking is don't include any CSS in the controls at let the component using them implement as needed. This is how it works on navigation, the justify controls takes an onChange function which is then used to insert the classes where needed.

Maybe it's better this way yes. We can revisit if needed, when applying this control to other blocks.

jasmussen commented:
We could write a "best practices" and suggest some good defaults, so the inconsistent markup behavior doesn't grow out of hands, in case we want to revisit?

mkaz committed:
Updates for CSS and usage clarification

- Simplifies CSS for toolbar to be generic, not just ul tag.
- Adds CSS back to navigation block since it is needed for > ul
- Updated documentation for using controls

mkaz commented:
Thanks, I've updated the component and I think it is ready.

Joen, good suggestion on the documentation, I updated the readme to I hope clarify. 

I'm keeping the CSS with the component, this will help standardize people using the same classnames when used. It will be left to the consumer of the controls to place the classname where needed, and we can revisit after seeing how it used and adjust.

ntsekouras reviewed (approved):
Thanks for working on this! I've left a couple of minor comments but this works well.

`Buttons` at least will need a `deprecation` when we'll integrate, but seems to provide a base component to use for other new components.

mkaz committed:
Add onChange as required

ntsekouras commented on packages/block-editor/src/components/justify-toolbar/index.js:
```diff
@@ -0,0 +1,78 @@
+/**
```
This comments about `External dependencies` should be removed.

mkaz replied:
Removed in 3dab628968

mkaz committed:
Deconstruct props

mkaz commented on packages/block-editor/src/components/justify-toolbar/index.js:
```diff
@@ -0,0 +1,78 @@
+};
+
+export function JustifyToolbar( props ) {
+	const {
```
Updated. :+1:

mkaz commented on packages/block-editor/src/components/justify-toolbar/README.md:
```diff
@@ -0,0 +1,81 @@
+
+
+#### `onChange`
+* **Type:** `Function`
```
Updated in dfc0c29

mkaz committed:
Remove extraneous external dependencies comment

youknowriad commented on packages/block-editor/src/components/justify-toolbar/index.js:
```diff
@@ -0,0 +1,72 @@
+	'space-between': justifySpaceBetween,
+};
+
+export function JustifyToolbar( {
```
Is the name good enough for a public API? Is it consistent with other `*Toolbar` components? Maybe it is, just making sure it's considered carefully.
